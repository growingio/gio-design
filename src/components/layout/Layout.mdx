import { Subtitle, ArgsTable } from '@storybook/addon-docs/blocks';
import Layout from './index';

# Layout 布局

<Subtitle>协助进行页面级整体布局。</Subtitle>

## 参数说明

<ArgsTable of={Layout} />

## 代码演示

### 样式 - 无侧导航页面
[Example](/?path=/story/basic-components-layout--default)

### 样式 - 收缩侧导航页面

[Example](/?path=/story/basic-components-layout--sider)

### 样式 - 悬浮式展开侧导航

[Example](/?path=/story/basic-components-layout--suspend)

### 样式 - 嵌入式展开侧导航

[Example](/?path=/story/basic-components-layout--embed)
