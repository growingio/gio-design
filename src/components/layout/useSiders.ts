import { useState, useMemo, useCallback } from 'react';
import { SiderState } from './interfaces';

const useSiders = (): [
  SiderState[],
  number,
  (siderId: string) => void,
  (sider: SiderState) => void,
  [number, number]
] => {
  const [siders, setSiders] = useState<SiderState[]>([]);

  const removeSider = (siderId: string) => {
    setSiders((_siders) => _siders.filter((_sider) => _sider.id !== siderId));
  };

  const updateSiders = useCallback((incomingSider: SiderState) => {
    setSiders((_siders) => {
      return [..._siders.filter((_sider) => _sider.id !== incomingSider.id), incomingSider];
    });
  }, []);

  const sidersWidth = useMemo(() => siders.reduce((prev: number, current: SiderState) => prev + current.width, 0), [
    siders,
  ]);

  const margin: [number, number] = useMemo(
    () => [
      siders.find((sider) => sider.suspendedPosition === 'left')?.collapsedWidth ?? 0,
      siders.find((sider) => sider.suspendedPosition === 'right')?.collapsedWidth ?? 0,
    ],
    [siders]
  );

  return [siders, sidersWidth, removeSider, updateSiders, margin];
};

export default useSiders;
