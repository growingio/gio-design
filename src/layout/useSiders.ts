import { useState, useMemo, useCallback } from 'react';
import { SiderState } from './interfaces';

const useSiders = (): [SiderState[], (siderId: string) => void, (sider: SiderState) => void, [number, number]] => {
  const [siders, setSiders] = useState<SiderState[]>([]);

  const removeSider = (siderId: string) => {
    setSiders((_siders) => _siders.filter((_sider) => _sider.id !== siderId));
  };

  const updateSiders = useCallback((incomingSider: SiderState) => {
    setSiders((_siders) => [..._siders.filter((_sider) => _sider.id !== incomingSider.id), incomingSider]);
  }, []);

  const margin: [number, number] = useMemo(
    () => [
      siders.find((sider) => sider.suspendedPosition === 'left')?.collapsedWidth ?? 0,
      siders.find((sider) => sider.suspendedPosition === 'right')?.collapsedWidth ?? 0,
    ],
    [siders]
  );

  return [siders, removeSider, updateSiders, margin];
};

export default useSiders;
