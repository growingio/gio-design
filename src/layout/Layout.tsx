import React, { useRef, createContext, useEffect, useContext } from 'react';
import classNames from 'classnames';
import { useSetState } from 'react-use';
import { usePrefixCls } from '@gio-design/utils';
import { LayoutProps, LayoutState, ContentState, LayoutContextType } from './interfaces';
import Header from './Header';
import Content from './Content';
import Sider from './Sider';
import useSiders from './useSiders';

const initLayoutState = { fixed: false };
const initContentState = { maxWidth: 0, margin: 0 };

export const LayoutContext = createContext<LayoutContextType>({
  layoutState: initLayoutState,
  contentState: initContentState,
} as LayoutContextType);

const Layout = ({ prefixCls: customizePrefixCls, fixed, className, style, children }: LayoutProps) => {
  const containerRef = useRef<HTMLDivElement>(null);
  const prefixCls = usePrefixCls('layout', customizePrefixCls);
  const [localLayoutState, setLayoutState] = useSetState<LayoutState>(initLayoutState);
  const [localContentState, setContentState] = useSetState<ContentState>(initContentState);
  const [siders, removeSider, updateSiders, margin] = useSiders();
  const { layoutState: parentLayoutState } = useContext(LayoutContext);

  useEffect(() => {
    setLayoutState({ fixed });
  }, [fixed, setLayoutState]);

  const mergedStyle = {
    ...{ marginLeft: margin[0], marginRight: margin[1] },
    ...(parentLayoutState.fixed ? { height: 'calc(100% - 60px)' } : {}),
    ...style,
  };

  return (
    <LayoutContext.Provider
      value={{
        contentState: localContentState,
        layoutState: localLayoutState,
        setLayoutState,
        setContentState,
        removeSider,
        updateSiders,
      }}
    >
      <section
        ref={containerRef}
        className={classNames(prefixCls, className, {
          [`${prefixCls}-has-sider`]: siders.length > 0,
          [`${prefixCls}-fixed`]: fixed,
        })}
        style={mergedStyle}
      >
        {children}
      </section>
    </LayoutContext.Provider>
  );
};

Layout.Header = Header;
Layout.Content = Content;
Layout.Sider = Sider;

export default Layout;
