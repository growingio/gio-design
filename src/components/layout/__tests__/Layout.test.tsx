import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { renderHook, act } from '@testing-library/react-hooks';
import Layout from '../layout';
import useSiders from '../useSiders';

describe('Testing Layout', () => {
  it('should be stable', () => {
    const { asFragment } = render(
      <Layout>
        <Layout.Sider defaultCollapsed />
        <Layout>
          <Layout.Header>
            <Layout.Header.HeaderSection />
            <Layout.Header.HeaderDivider />
          </Layout.Header>
          <Layout.Content />
        </Layout>
      </Layout>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('Content maxWidth', () => {
    const { getByText } = render(
      <Layout>
        <Layout.Content maxWidth="auto">content</Layout.Content>
      </Layout>
    );
    const contentElement = getByText('content');
    expect(contentElement.style.getPropertyValue('margin')).toBe('0px 20px');
  });

  test('Content margin', () => {
    const { getByText, rerender } = render(
      <Layout>
        <Layout.Content maxWidth="auto" margin={10}>
          content
        </Layout.Content>
      </Layout>
    );
    expect(getByText('content').style.getPropertyValue('margin')).toBe('0px 10px');
    Object.defineProperty(HTMLElement.prototype, 'getBoundingClientRect', {
      configurable: true,
      value: () => ({ width: 2000 }),
    });
    rerender(
      <Layout style={{ width: 2000 }}>
        <Layout.Content>content</Layout.Content>
      </Layout>
    );
    expect(getByText('content').style.getPropertyValue('margin')).toBe('0px auto');
  });

  test('sider trigger', () => {
    const { getByText, getByTestId, rerender } = render(
      <Layout>
        <Layout.Sider trigger={<div data-testid="trigger-test">trigger</div>} />
      </Layout>
    );
    expect(getByText('trigger').getAttribute('data-testid')).toBe('trigger-test');
    expect(getByTestId('sider-content').nextElementSibling).toBe(getByText('trigger'));

    rerender(
      <Layout>
        <Layout.Sider trigger="bottom" />
      </Layout>
    );

    expect(getByTestId('sider-content').nextElementSibling?.className.includes('gio-layout-sider-bottom-trigger')).toBe(
      true
    );
    expect(getByTestId('right-icon')).toBeTruthy();
    fireEvent(
      getByTestId('sider-content').nextElementSibling as Element,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    );
    expect(getByTestId('left-icon')).toBeTruthy();

    rerender(
      <Layout>
        <Layout.Sider trigger="top" />
      </Layout>
    );
    expect(getByTestId('sider-content').nextElementSibling).toBe(null);
  });

  test('useSiders', () => {
    const { result } = renderHook(() => useSiders());
    const _updateSiders = result.current[3];
    act(() => {
      _updateSiders({ id: '1', width: 200, collapsedWidth: 80, suspendedPosition: 'left' });
      _updateSiders({ id: '2', width: 300, collapsedWidth: 100, suspendedPosition: 'right' });
      _updateSiders({ id: '3', width: 100, collapsedWidth: 50 });
    });
    const [siders, sidersWidth, removeSider, updateSiders, margin] = result.current;
    expect(siders.length).toBe(3);
    expect(sidersWidth).toBe(600);
    expect(margin).toStrictEqual([80, 100]);
    act(() => {
      removeSider('1');
    });
    expect(result.current[0].length).toBe(2);
    expect(result.current[1]).toBe(400);
    expect(result.current[4]).toStrictEqual([0, 100]);
    act(() => {
      updateSiders({ id: '3', width: 100, collapsedWidth: 50, suspendedPosition: 'left' });
    });
    expect(result.current[4]).toStrictEqual([50, 100]);
    act(() => {
      removeSider('2');
    });
    expect(result.current[4]).toStrictEqual([50, 0]);
  });
});
