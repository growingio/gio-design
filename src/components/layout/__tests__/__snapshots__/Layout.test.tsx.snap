// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Layout should be stable 1`] = `
<DocumentFragment>
  <section
    class="gio-layout gio-layout-has-sider"
    style="margin-left: 0px; margin-right: 0px;"
  >
    <aside
      class="gio-layout-sider gio-layout-sider-collapsed"
      style="--layout-sider-width: 200px; --layout-sider-collapsedWidth: 80px; --layout-sider-collapsedWidth-negative: -80px;"
    >
      <div
        class="gio-layout-sider-content"
        data-testid="sider-content"
      />
    </aside>
    <section
      class="gio-layout"
      style="margin-left: 0px; margin-right: 0px;"
    >
      <header
        class="gio-layout-header"
      >
        <section
          class="gio-layout-header-section"
          style="--layout-header-section-justify: space-around;"
        />
        <div
          class="gio-layout-header-divider"
        />
      </header>
      <main
        class="gio-layout-content"
        style="flex-grow: 1; margin: 0px 20px;"
      />
    </section>
  </section>
</DocumentFragment>
`;
